### PR TITLE
feat: demo wallet

### LINKED ISSUE #895: Allow use of app without connected wallet/"demo mode"
# Overview

Many users have requested to be able to use the app (app.shapeshift.com) without first connecting a wallet. This makes sense from an engagement perspective - people want to see what things look and feel like before they use it.

# Reference

- instantiate a demo wallet like shown https://github.com/shapeshift/hdwallet/pull/452/commits/e59e755b6403cdfd2bbb331df2a0f535eef00a31#diff-673a1350d3a18db303ae7b7494c175c13e74e0dbd8faef3b1200d9708b028f9aR334-R335
- make an initial PR to update the `useWallet` hooks to provide an `isDemoWallet` boolean

# Guidance

At a high level, the app should be as functional as possible, without the user being able to get into a broken state.

- Use the `useWallet` hook to detect the presence (or rather, lack thereof) of a wallet for logic
- Add a grey button to the splash screen below the Connect Wallet, saying "Try without a wallet..."
- Redirect to [Bitcoin asset page](https://app.shapeshift.com/assets/bitcoin) as the first screen instead of `/dashboard`
- Disable send/receive buttons on asset pages
- Hide account page link in nav
- Hide accounts from asset pages
- Hide balance chart buttons, make sure pages default to price charts
- Disconnect wallet button should disconnect wallet and clear portfolio, but not boot the user back out to the `connect-wallet` screen
- Trade dialog
  - Sell assets should be fetched from all assets list, not users portfolio assets
  - Confirm trade button should be disabled
- Defi vaults - show modal but stop the user at step 1

# Acceptance criteria

- No regressions across the app
- App is usable to the fullest extent possible without connecting a wallet

# Bounty Hunters 
- Join our discord [here](https://discord.gg/shapeshift). Once you join you can see more about this issue [here](https://discord.com/channels/554694662431178782/873244924026449940/925799058537336853)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info

### LINKED ISSUE #1621: Demo Wallet - Allow use of app without connected wallet/"demo mode" (v2)
# Overview

Many users have requested to be able to use the app (app.shapeshift.com) without first connecting a wallet. This makes sense from an engagement perspective - people want to see what things look and feel like before they use it.

**Objective**

*As a user I want to be able to view a demo of the Shapeshift web app before I connect my wallet.*

*[app.shapeshift.com](http://app.shapeshift.com/)*

1. On the splash screen remove the arrow from the “connect wallet” button 

2. User will see a link under the connect wallet button for “or view a demo —>” [mock up](https://invis.io/DT12C1FLC7QA#/463792391_Artboard_Copy)
    a. This will launch a “native wallet”

3. “view a demo” will open a Demo wallet created by the Operations Workstream 

4. Whenever a user clicks an action that will require a wallet (Send, Receive, Trade, Earn), they see the “connect wallet” modal

5. The default timeframe on all charts should be updated to `Timeframe.MONTH` (for all wallet types) 

### References and additional details

Product specification
- [The most up-to-date Product Spec](https://shapeshift.notion.site/Demo-Wallet-Spec-2509a0b0fe034ca6a477b7b4b31740d2)

Related issues

- [Create xpub list for official demo wallet](https://github.com/shapeshift/security/issues/30)
- [Feature: Create a "demo" wallet type](https://github.com/shapeshift/hdwallet/issues/445)
- [Allow use of app without connected wallet/"demo mode"](https://github.com/shapeshift/web/issues/895)
- [The original issue](https://github.com/shapeshift/web/issues/895)

### Acceptance Criteria

- A user is able to view a demo of the ShapeShift web app before they connect a wallet.

- No regressions across the app

- App is usable to the fullest extent possible without connecting a wallet

- Interacting with the app in a way that requires a connected wallet should prompt the user to "Connect A Wallet"

### Need By Date

N/A

### Screenshots/Mockups

N/A

### PR DESCRIPTION
**Pull request by 0xf11nix**
## Description

This PR implements the Demo Wallet feature, which is a loose combination and variation on the following specifications:

- [Related issue](https://github.com/shapeshift/web/issues/1621)
- [Product specification](https://shapeshift.notion.site/Demo-Wallet-Spec-2509a0b0fe034ca6a477b7b4b31740d2)

### Note: 
From the [current spec](https://shapeshift.notion.site/Demo-Wallet-Spec-2509a0b0fe034ca6a477b7b4b31740d2), using the `type` property of the useWallet hook is not feasible, since this property is mutable in the context of the "actual connected wallet" and is currently (mostly) used to determine the initial state of the "Connect A Wallet" modal.

As a result, and with some valuable insight from @majorhayes (much appreciated), it was decided that using the existing `isConnected` property would cover almost all scenarios where we need to differentiate between an "actually connected" wallet, and a wallet that is not connected, or a demo wallet. The few exceptions for this could also be accounted for by checking the `deviceId` in walletInfo`.
This also has the added benefit that any future functionality requiring a wallet to be "connected" is less likely to break the feature if someone forgets to add the specific flag to check for a demo wallet.

Still outstanding:

- ~~Production Demo Wallet xPubs: to be created and added~~ 
- ~~E2E tests: underway and will push and update soon~~ - to be included as part of (or immediately following) the PR that removes the feature flag.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Resolves #1621 

## Risk

The risk to production functionality for this feature is relatively low, since most of the changes effect only the the interaction with the platform by a user attempting to "demo" functionality, without a wallet connected.

## Testing

There should be no regression in existing behaviour, either when a wallet is connected, or when in "Demo Mode".
When in "Demo Mode", the app should be as usable as possible without the need to connect a wallet. The user should be able to progress with functionality, such a making a trade or staking an asset, but only to the point where an "actual" wallet is required. At this point, they should be prompted to "Connect A Wallet".

These changes are currently behind a feature flag, and a more extensive test plan will be included on the PR that remove the flag for Production.

## Screenshots (if applicable)

Screen Shot 2022-05-02 at 10 37 04 pm

### CODE REVIEW

**Comment by 0xdef1cafe** [comment]
suggestion for @DiggyDiggy2 @reallybeard @SS-FOX-McCloud - if demo wallet is connect, we change the default timeframe on the chart to the all timeframe - otherwise we land on a zero balance chart. thoughts?

**Comment by reallybeard** [comment]
> suggestion for @DiggyDiggy2 @reallybeard @SS-FOX-McCloud - if demo wallet is connect, we change the default timeframe on the chart to the all timeframe - otherwise we land on a zero balance chart. thoughts?

I'd say that is fine, better than showing them a flat line.

**Comment by DiggyDiggy2** [comment]
can we get a demo wallet connected here that has assets in it for testing? is this the one that Ops made?

**Comment by 0xdef1cafe** [comment]
@0xf11nix w.r.t this comment

```
Still outstanding:

Production Demo Wallet xPubs: to be created and added
E2E tests: underway and will push and update soon
```

does this open hdwallet PR address it? https://github.com/shapeshift/hdwallet/pull/519

**Comment by 0xf11nix** [comment]
> does this open hdwallet PR address it? https://github.com/shapeshift/hdwallet/pull/519

@0xdef1cafe I believe this is the issue related to generating the list of xpubs for the "official" demo wallet
https://github.com/shapeshift/security/issues/30

@mrnerdhair Is this still the case?

File: src/components/Layout/Header/NavBar/UserMenu.tsx [hunk 1]
```diff
@@ -98,7 +98,15 @@ const WalletButton: FC<WalletButtonProps> = ({
       isLoading={isLoadingLocalWallet}
       leftIcon={
         <HStack>
-          {isConnected ? undefined : <WarningTwoIcon ml={2} w={3} h={3} color='yellow.500' />}
+          {isConnected || walletInfo?.deviceId === 'DemoWallet' ? undefined : (
```
Comment (0xApotheosis) [inline src/components/Layout/Header/NavBar/UserMenu.tsx:101]: An observation: I'm not sure which is correct, but the existing pattern in the codebase is to return `null`, rather than `undefined`, from a conditional in JSX.

That said, I can't find any documentation asserting that `null` is the "right" way to do it, and `undefined` feels more right to me.

File: src/assets/translations/main.json [hunk 2]
```diff
@@ -109,7 +109,8 @@
     "defiUniverse": "DeFi Universe",
     "body": "100% Free. Open Source. Community Owned. Non-Custodial. Private.",
     "header": "Please connect a wallet to get started",
-    "cta": "Connect Wallet"
+    "cta": "Connect Wallet",
+    "orViewADemo": "or view a Demo"
```
Comment (0xApotheosis) [inline src/assets/translations/main.json:113]: `demo` is lower case in the mockups.

```suggestion
    "orViewADemo": "or view a demo"
```

**Review by 0xApotheosis — DISMISSED** [review]
LGTM, works well - couldn't break it in testing.

2 small things, one of them more of an observation.

**Review by 0xApotheosis — COMMENTED** [review]
Meant to leave a "Comment", not sure if this update will undo the approval...
In any case, let's leave in draft until E2E tests are in, and production-ready xpubs are ready.

File: src/components/Layout/Header/NavBar/UserMenu.tsx (hunk 1 above)
Comment (gomesalexandre) [reply src/components/Layout/Header/NavBar/UserMenu.tsx:101]: By **convention**, short-circuiting JSX renders in React with ternaries is done with `null`, with the happy render path first: 
```suggestion
          {!isConnected || walletInfo?.deviceId !== 'DemoWallet' ? happyJsxPath : null
```

In theory, react handles `false`, `null` and `undefined` the same way however: https://reactjs.org/docs/jsx-in-depth.html#booleans-null-and-undefined-are-ignored
> false, null, undefined, and true are valid children. They simply don’t render.

It means that this can be simplified with an expression such as (pseudo-code written in the comment box, obviously untested):

```suggestion
          {(!isConnected || walletInfo?.deviceId !== 'DemoWallet') ?? happyJsxPath}
```

**Comment by mrnerdhair** [comment]
> does this open hdwallet PR address it? [shapeshift/hdwallet#519](https://github.com/shapeshift/hdwallet/pull/519)

It does now... shoved the code to generate the xpub list in there, and put the demo wallet's xpub list in a [comment](https://github.com/shapeshift/hdwallet/pull/519#issuecomment-1111496932)

File: src/constants/PublicWalletXpubs.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// FIXME: @0Xf11nix - Do not commit to Prod, replace with actual creds
```
Comment (0xdef1cafe) [inline src/constants/PublicWalletXpubs.ts:1]: heyo - please replace with the demo wallet xpubs in the comment here

https://github.com/shapeshift/hdwallet/pull/519#issuecomment-1111496932

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
@0xf11nix few things here

1. let's get the updated demo wallet xpubs committed
2. let's set the default timeframe on the charts to `Timeframe.All` if the demo wallet is connected
3. can we add a feature flag so we can get this merged/iterate without a goat?

File: src/components/Layout/Header/NavBar/UserMenu.tsx (hunk 1 above)
Comment (0xApotheosis) [reply src/components/Layout/Header/NavBar/UserMenu.tsx:101]: TIL @gomesalexandre  👍

**Comment by 0xf11nix** [comment]
> A few things I've found, maybe because this is still a draft and it's being tackled on but leaving it here just as a reference as a non-aggressive comment:
> 
> * It's possible to click the "Receive" button on account/asset page - which brings the "Connect Wallet" Modal
> * Accounts link is still on the nav
> * Accounts link is still on the asset page
> * Balance button is still there
> * Disconnect button still takes back to the "Connect Wallet" splash screen

@gomesalexandre Thank you for your review.
Unfortunately, there was a little misunderstanding between the [original github issue](https://github.com/shapeshift/web/issues/895) and the [most recent product specification](https://projects.invisionapp.com/share/DT12C1FLC7QA#/screens/463792391_Artboard_Copy). After discussing with @DiggyDiggy2 and @0xdef1cafe, we settled on something much closer to the latter. 
To try and clear things up, I have created [a new issue](https://github.com/shapeshift/web/issues/1621) to try and better encapsulate the actual desired outcome. Note: the previously hidden elements, as well as forcing the using back to the "connect wallet" screen, are no longer required as part of this new spec.
Again, thanks for taking the time to review these changes, I do hope this new spec better explains some of the design decision taken.

File: src/constants/PublicWalletXpubs.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/constants/PublicWalletXpubs.ts]: Addressed in https://github.com/shapeshift/web/pull/1585/commits/94d0ea775f946cb294188c3a45c2e6a86109e075

**Comment by 0xf11nix** [comment]
> @0xf11nix few things here
> 
> 1. let's get the updated demo wallet xpubs committed
> 2. let's set the default timeframe on the charts to `Timeframe.All` if the demo wallet is connected
> 3. can we add a feature flag so we can get this merged/iterate without a goat?

@0xdef1cafe I have addressed these requested changes in the following commits:

1. 94d0ea775f946cb294188c3a45c2e6a86109e075
2. cc9e95a44ab3493b752be5adce17305d7d255b13
3. 6fb2555bb63af63abd3b89e65bb8f145deecef3d

The feature flag might be a little annoying for testing, since you need to connect a wallet in order to first enable the Demo Wallet flag, but I guess we can work around that.

Now that the feature is behind a flag, my Cypress tests are a bit broken, so I'll push those once I find a work-around.

**Comment by 0xdef1cafe** [comment]
@0xf11nix this looks dope AF. i've turned the flag on in preview cloudflare environments and will do the same for fleek.

two things i noticed while testing this with @DiggyDiggy2 

* **Bug fix - defi/earn deposit and withdraw - the "Sign & Broadcast" button should open the connect wallet modal**
* **Bug fix - "Confirm and Trade" button on trade widget white screens then refreshes page to `/?#/dashboard`**
* Let's change the default timeframe to 1 month on the portfolio charts - so we're not seeing a long period of flatlined zero balance

let's get these small fixes made and this PR merged behind the flag, and we're probably good to turn it on, after another mini goat.

additionally, we created another ticket that product will fully flesh out for some additional enhancements. this is scope creep but all really valuable stuff

https://github.com/shapeshift/web/issues/1637

**Comment by 0xf11nix** [comment]
> - Bug fix - defi/earn deposit and withdraw - the "Sign & Broadcast" button should open the connect wallet modal
> - Bug fix - "Confirm and Trade" button on trade widget white screens then refreshes page to /?#/dashboard
> - Let's change the default timeframe to 1 month on the portfolio charts - so we're not seeing a long period of flatlined zero balance. **let's make this

@0xdef1cafe I have addressed the three issues above.

After getting the wallet topped up with some ETH, we can now properly test some of the edge cases.
There does appear to be a new issue related to the approval step, before "Confirm and trade", so I'm working on that now.

I will also address the merge conflict before I mark these changes as ready for review.

**Comment by 0xdef1cafe** [comment]
> > * Bug fix - defi/earn deposit and withdraw - the "Sign & Broadcast" button should open the connect wallet modal
> > * Bug fix - "Confirm and Trade" button on trade widget white screens then refreshes page to /?#/dashboard
> > * Let's change the default timeframe to 1 month on the portfolio charts - so we're not seeing a long period of flatlined zero balance. **let's make this
> 
> @0xdef1cafe I have addressed the three issues above.
> 
> After getting the wallet topped up with some ETH, we can now properly test some of the edge cases. There does appear to be a new issue related to the approval step, before "Confirm and trade", so I'm working on that now.
> 
> I will also address the merge conflict before I mark these changes as ready for review.

awesome. also when i was reviewing with @DiggyDiggy2 in person - let's just change the default timeframe for all charts to 1 month, day is too noisy, month gives enough context, and will have a full history for this demo wallet.

**Comment by 0xf11nix** [comment]
> awesome. also when i was reviewing with @DiggyDiggy2 in person - let's just change the default timeframe for all charts to 1 month, day is too noisy, month gives enough context, and will have a full history for this demo wallet.

@0xdef1cafe I think we should be good to go now. I have made the above change, as requested, and addressed a number of other small bugs that were raised over the last couple of days.

As per our discussion, and noted in the description above, we will hold off on the E2E tests for now and include them in a follow-up PR once the feature flag is removed.
